File: frontend/js/pages/receptionist.js
// ===== Receptionist Dashboard =====
async function renderReceptionistDashboard() {
  const data = await api.get('/receptionist/dashboard');
  if (!data) return '<div class="spinner"></div>';
  const s = data.stats;
  return `<div class="page-header"><h1>Welcome, ${data.user.name} 🏥</h1><p>Front desk management</p></div>
    <div class="stats-grid">
      ${statCard('📅','Today\'s Appointments', s.today_appointments, '#e8795a')}
      ${statCard('👥','Total Patients', s.total_patients, '#3b82f6')}
      ${statCard('👨‍⚕️','Doctors Available', s.total_doctors, '#0d9488')}
      ${statCard('⏳','Pending', s.pending_appointments, '#f59e0b')}
    </div>
    <div class="grid-2">
      <div class="card"><div class="card-header"><h3>📋 Recent Appointments</h3></div><div class="card-body">
        ${data.recent_appointments.length ? data.recent_appointments.map(a => `<div class="alert-item">
          <div class="alert-dot" style="background:${a.status==='pending'?'var(--accent-amber)':a.status==='confirmed'?'var(--accent-emerald)':'var(--accent-blue)'}"></div>
          <div style="flex:1"><strong>${a.patient_name}</strong> → ${a.doctor_name}<br>
          <span style="font-size:0.8rem;color:var(--text-secondary)">${a.date} at ${a.time} · ${a.specialization}</span></div>
          <span class="badge ${a.status==='pending'?'badge-warning':a.status==='confirmed'?'badge-success':'badge-info'}">${a.status.toUpperCase()}</span>
        </div>`).join('') : '<p style="color:var(--text-muted)">No recent appointments.</p>'}
      </div></div>
      <div class="card"><div class="card-header"><h3>⚡ Quick Actions</h3></div><div class="card-body" style="display:flex;flex-direction:column;gap:10px">
        <button class="btn btn-primary" onclick="navigateTo('rec-register')" style="width:100%;background:linear-gradient(135deg,#e8795a,#f59e0b)">👤 Register New Patient</button>
        <button class="btn btn-outline" onclick="navigateTo('rec-book')" style="width:100%">📅 Book Appointment</button>
        <button class="btn btn-outline" onclick="navigateTo('rec-patients')" style="width:100%">👥 View All Patients</button>
        <button class="btn btn-outline" onclick="navigateTo('rec-appointments')" style="width:100%">📋 View All Appointments</button>
      </div></div>
    </div>`;
}

// ===== Register Patient =====
function renderRecRegister() {
  return `<div class="page-header"><h1>👤 Register New Patient</h1><p>Add a new patient to the system</p></div>
    <div class="card"><div class="card-body">
      <div class="form-row"><div class="form-group"><label>Full Name *</label><input type="text" id="rpName" placeholder="Patient full name" required></div><div class="form-group"><label>Email *</label><input type="email" id="rpEmail" placeholder="patient@email.com" required></div></div>
      <div class="form-row"><div class="form-group"><label>Phone</label><input type="text" id="rpPhone" placeholder="9876543210"></div><div class="form-group"><label>Age</label><input type="number" id="rpAge" placeholder="30"></div></div>
      <div class="form-row"><div class="form-group"><label>Gender</label><select id="rpGender"><option value="">Select</option><option>Male</option><option>Female</option><option>Other</option></select></div><div class="form-group"><label>Blood Group</label><select id="rpBlood"><option value="">Select</option><option>A+</option><option>A-</option><option>B+</option><option>B-</option><option>AB+</option><option>AB-</option><option>O+</option><option>O-</option></select></div></div>
      <div class="form-row"><div class="form-group"><label>Height (cm)</label><input type="number" id="rpHeight" placeholder="170"></div><div class="form-group"><label>Weight (kg)</label><input type="number" id="rpWeight" placeholder="70"></div></div>
      <div class="form-group"><label>Medical History</label><textarea id="rpHistory" placeholder="Previous conditions, surgeries..."></textarea></div>
      <div class="form-row"><div class="form-group"><label>Allergies</label><input type="text" id="rpAllergies" placeholder="e.g., Penicillin"></div><div class="form-group"><label>Emergency Contact</label><input type="text" id="rpEmergency" placeholder="Contact number"></div></div>
      <div class="form-group"><label>Address</label><textarea id="rpAddress" placeholder="Full address"></textarea></div>
      <button class="btn btn-primary" onclick="submitRegisterPatient()" style="width:100%;margin-top:8px">✅ Register Patient</button>
      <div id="rpResult" style="margin-top:16px"></div>
    </div></div>`;
}

async function submitRegisterPatient() {
  const name = document.getElementById('rpName').value.trim();
  const email = document.getElementById('rpEmail').value.trim();
  if (!name || !email) return showToast('Name and email are required', 'error');

  const payload = {
    full_name: name,
    email: email,
    phone: document.getElementById('rpPhone').value,
    age: document.getElementById('rpAge').value ? +document.getElementById('rpAge').value : null,
    gender: document.getElementById('rpGender').value || null,
    blood_group: document.getElementById('rpBlood').value || null,
    height_cm: document.getElementById('rpHeight').value ? +document.getElementById('rpHeight').value : null,
    weight_kg: document.getElementById('rpWeight').value ? +document.getElementById('rpWeight').value : null,
    medical_history: document.getElementById('rpHistory').value,
    allergies: document.getElementById('rpAllergies').value,
    emergency_contact: document.getElementById('rpEmergency').value,
    address: document.getElementById('rpAddress').value,
  };

  try {
    const result = await api.post('/receptionist/register-patient', payload);
    document.getElementById('rpResult').innerHTML = `<div class="card" style="border-color:rgba(16,185,129,0.3)"><div class="card-body">
      <h3>✅ ${result.message}</h3>
      <p style="margin-top:8px"><strong>Patient ID:</strong> ${result.patient_id}</p>
      <p><strong>Email:</strong> ${result.email}</p>
      <p><strong>Default Password:</strong> <code style="background:#f1f5f9;padding:2px 8px;border-radius:4px">${result.default_password}</code></p>
    </div></div>`;
    showToast('Patient registered successfully!');
  } catch (err) { showToast(err.message, 'error'); }
}

// ===== Book Appointment =====
async function renderRecBook() {
  let doctorsHtml = '';
  let patientsHtml = '';
  try {
    const doctors = await api.get('/receptionist/doctors');
    doctorsHtml = doctors.map(d => `<option value="${d.id}">${d.name} — ${d.specialization} (⭐${d.rating})</option>`).join('');
  } catch { doctorsHtml = '<option>Error loading doctors</option>'; }

  try {
    const patients = await api.get('/receptionist/patients');
    patientsHtml = patients.map(p => `<option value="${p.id}">${p.name} — ${p.email}${p.blood_group ? ' ('+p.blood_group+')' : ''}</option>`).join('');
  } catch { patientsHtml = '<option>Error loading patients</option>'; }

  const tomorrow = new Date(Date.now()+86400000).toISOString().split('T')[0];

  return `<div class="page-header"><h1>📅 Book Appointment</h1><p>Assign a patient to a doctor</p></div>
    <div class="card"><div class="card-body">
      <div class="form-group"><label>Select Patient *</label><select id="baPatient">${patientsHtml}</select></div>
      <div class="form-group"><label>Select Doctor *</label><select id="baDoctor">${doctorsHtml}</select></div>
      <div class="form-row"><div class="form-group"><label>Date *</label><input type="date" id="baDate" value="${tomorrow}"></div><div class="form-group"><label>Time *</label><input type="time" id="baTime" value="10:00"></div></div>
      <div class="form-group"><label>Reason for Visit</label><textarea id="baReason" placeholder="Symptoms, check-up, follow-up..."></textarea></div>
      <button class="btn btn-primary" onclick="submitBookAppointment()" style="width:100%;margin-top:8px">📅 Book Appointment</button>
      <div id="baResult" style="margin-top:16px"></div>
    </div></div>`;
}

async function submitBookAppointment() {
  const patientId = document.getElementById('baPatient').value;
  const doctorId = document.getElementById('baDoctor').value;
  const date = document.getElementById('baDate').value;
  const time = document.getElementById('baTime').value;
  if (!patientId || !doctorId || !date) return showToast('Patient, doctor, and date are required', 'error');

  try {
    const result = await api.post('/receptionist/appointments', {
      patient_id: +patientId,
      doctor_id: +doctorId,
      date: date,
      time: time || '10:00',
      reason: document.getElementById('baReason').value,
    });
    document.getElementById('baResult').innerHTML = `<div class="card" style="border-color:rgba(16,185,129,0.3)"><div class="card-body"><h3>✅ ${result.message}</h3><p style="margin-top:4px">Appointment ID: ${result.appointment_id}</p></div></div>`;
    showToast('Appointment booked!');
  } catch (err) { showToast(err.message, 'error'); }
}

// ===== All Patients =====
async function renderRecPatients() {
  let html = `<div class="page-header"><h1>👥 All Patients</h1><p>Registered patients in the system</p></div>
    <div style="display:flex;gap:12px;margin-bottom:20px"><input type="text" id="recPatSearch" placeholder="Search by name..." style="flex:1;padding:12px 16px;background:#f8fafc;border:1px solid var(--border-light);border-radius:10px;color:var(--text-primary);outline:none" oninput="searchRecPatients()"><button class="btn btn-primary" onclick="navigateTo('rec-register')">+ Add Patient</button></div>`;
  try {
    const patients = await api.get('/receptionist/patients');
    window._recPatients = patients;
    html += `<div class="card"><div class="card-body"><div class="table-wrap"><table><thead><tr><th>ID</th><th>Name</th><th>Email</th><th>Phone</th><th>Age</th><th>Gender</th><th>Blood</th><th>Action</th></tr></thead><tbody id="recPatTable">${renderRecPatRows(patients)}</tbody></table></div></div></div>`;
  } catch { html += '<p>Failed to load patients.</p>'; }
  return html;
}

function renderRecPatRows(patients) {
  return patients.map(p => `<tr><td>${p.id}</td><td><strong>${p.name}</strong></td><td>${p.email}</td><td>${p.phone||'—'}</td><td>${p.age||'—'}</td><td>${p.gender||'—'}</td><td>${p.blood_group||'—'}</td><td><button class="btn btn-sm btn-primary" onclick="quickBookForPatient(${p.id},'${p.name}')">Book →</button></td></tr>`).join('');
}

function searchRecPatients() {
  const q = document.getElementById('recPatSearch').value.toLowerCase();
  const filtered = (window._recPatients||[]).filter(p => p.name.toLowerCase().includes(q) || p.email.toLowerCase().includes(q));
  document.getElementById('recPatTable').innerHTML = renderRecPatRows(filtered);
}

function quickBookForPatient(id, name) {
  navigateTo('rec-book');
  setTimeout(() => {
    const sel = document.getElementById('baPatient');
    if (sel) sel.value = id;
  }, 300);
}

// ===== All Appointments =====
async function renderRecAppointments() {
  let html = `<div class="page-header"><h1>📋 All Appointments</h1><p>View and track all appointments</p></div>
    <div style="display:flex;gap:12px;margin-bottom:20px"><input type="date" id="recAptDate" style="padding:12px 16px;background:#f8fafc;border:1px solid var(--border-light);border-radius:10px;color:var(--text-primary);outline:none" onchange="filterRecAppointments()"><button class="btn btn-outline" onclick="document.getElementById('recAptDate').value='';filterRecAppointments()">Clear Filter</button><button class="btn btn-primary" onclick="navigateTo('rec-book')" style="margin-left:auto">+ New Appointment</button></div>`;
  try {
    const apts = await api.get('/receptionist/appointments');
    window._recAppointments = apts;
    html += `<div class="card"><div class="card-body"><div class="table-wrap"><table><thead><tr><th>ID</th><th>Patient</th><th>Doctor</th><th>Specialization</th><th>Date</th><th>Time</th><th>Reason</th><th>Status</th></tr></thead><tbody id="recAptTable">${renderRecAptRows(apts)}</tbody></table></div></div></div>`;
  } catch { html += '<p>Failed to load appointments.</p>'; }
  return html;
}

function renderRecAptRows(apts) {
  return apts.map(a => `<tr>
    <td>${a.id}</td><td><strong>${a.patient_name}</strong></td><td>${a.doctor_name}</td><td><span class="badge badge-info">${a.specialization||'—'}</span></td>
    <td>${a.date}</td><td>${a.time}</td><td>${a.reason||'—'}</td>
    <td>${a.status==='pending'?'<span class="badge badge-warning">PENDING</span>':a.status==='confirmed'?'<span class="badge badge-success">CONFIRMED</span>':a.status==='completed'?'<span class="badge badge-info">COMPLETED</span>':'<span class="badge badge-danger">CANCELLED</span>'}</td>
  </tr>`).join('');
}

async function filterRecAppointments() {
  const date = document.getElementById('recAptDate').value;
  try {
    const apts = await api.get('/receptionist/appointments' + (date ? `?date_filter=${date}` : ''));
    document.getElementById('recAptTable').innerHTML = renderRecAptRows(apts);
  } catch {}
}

File: backend/app/routers/receptionist.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func
from datetime import date
from app.database import get_db
from app.models.user import User, UserRole
from app.models.patient import PatientProfile
from app.models.doctor import DoctorProfile
from app.models.appointment import Appointment, AppointmentStatus
from app.schemas.schemas import ReceptionistPatientRegister, ReceptionistAppointmentCreate
from app.utils.security import get_current_user, require_role, hash_password

router = APIRouter(prefix="/api/receptionist", tags=["Receptionist Portal"])

# ── Dashboard ──────────────────────────────────────────
@router.get("/dashboard")
async def receptionist_dashboard(
    current_user: User = Depends(require_role(UserRole.RECEPTIONIST)),
    db: AsyncSession = Depends(get_db)
):
    today = date.today().isoformat()

    # Today's appointments
    result = await db.execute(
        select(func.count()).select_from(Appointment).where(Appointment.date == today)
    )
    today_appointments = result.scalar() or 0

    # Total patients
    result = await db.execute(
        select(func.count()).select_from(User).where(User.role == UserRole.PATIENT)
    )
    total_patients = result.scalar() or 0

    # Total doctors
    result = await db.execute(
        select(func.count()).select_from(User).where(User.role == UserRole.DOCTOR)
    )
    total_doctors = result.scalar() or 0

    # Pending appointments
    result = await db.execute(
        select(func.count()).select_from(Appointment).where(
            Appointment.status == AppointmentStatus.PENDING
        )
    )
    pending_appointments = result.scalar() or 0

    # Total appointments
    result = await db.execute(select(func.count()).select_from(Appointment))
    total_appointments = result.scalar() or 0

    # Recent appointments
    result = await db.execute(
        select(Appointment).order_by(Appointment.created_at.desc()).limit(5)
    )
    recent = result.scalars().all()

    return {
        "user": {"id": current_user.id, "name": current_user.full_name},
        "stats": {
            "today_appointments": today_appointments,
            "total_patients": total_patients,
            "total_doctors": total_doctors,
            "pending_appointments": pending_appointments,
            "total_appointments": total_appointments,
        },
        "recent_appointments": [{
            "id": a.id,
            "patient_name": a.patient_name,
            "doctor_name": a.doctor_name,
            "specialization": a.specialization,
            "date": a.date,
            "time": a.time,
            "status": a.status.value,
        } for a in recent]
    }


# ── List all doctors ────────────────────────────────────
@router.get("/doctors")
async def list_doctors(
    current_user: User = Depends(require_role(UserRole.RECEPTIONIST)),
    db: AsyncSession = Depends(get_db)
):
    result = await db.execute(
        select(DoctorProfile, User).join(User, DoctorProfile.user_id == User.id)
    )
    rows = result.all()

    return [{
        "id": user.id,
        "name": user.full_name,
        "specialization": doc.specialization,
        "experience_years": doc.experience_years,
        "qualification": doc.qualification,
        "rating": doc.rating,
        "consultation_fee": doc.consultation_fee,
        "is_available": doc.is_available,
        "available_hours": doc.available_hours,
    } for doc, user in rows]


# ── Register a new patient ──────────────────────────────
@router.post("/register-patient")
async def register_patient(
    data: ReceptionistPatientRegister,
    current_user: User = Depends(require_role(UserRole.RECEPTIONIST)),
    db: AsyncSession = Depends(get_db)
):
    # Check if email already registered
    result = await db.execute(select(User).where(User.email == data.email))
    if result.scalar_one_or_none():
        raise HTTPException(status_code=400, detail="Email already registered")

    # Create user with a default password
    user = User(
        email=data.email,
        password_hash=hash_password("password123"),
        full_name=data.full_name,
        role=UserRole.PATIENT,
        phone=data.phone or ""
    )
    db.add(user)
    await db.commit()
    await db.refresh(user)

    # Create patient profile
    profile = PatientProfile(
        user_id=user.id,
        age=data.age,
        gender=data.gender,
        blood_group=data.blood_group,
        height_cm=data.height_cm,
        weight_kg=data.weight_kg,
        medical_history=data.medical_history or "",
        allergies=data.allergies or "",
        emergency_contact=data.emergency_contact or "",
        address=data.address or "",
    )
    db.add(profile)
    await db.commit()

    return {
        "message": f"Patient {data.full_name} registered successfully",
        "patient_id": user.id,
        "email": user.email,
        "default_password": "password123"
    }


# ── List all patients ───────────────────────────────────
@router.get("/patients")
async def list_patients(
    q: str = "",
    current_user: User = Depends(require_role(UserRole.RECEPTIONIST)),
    db: AsyncSession = Depends(get_db)
):
    query = select(User, PatientProfile).outerjoin(
        PatientProfile, User.id == PatientProfile.user_id
    ).where(User.role == UserRole.PATIENT)

    if q:
        query = query.where(User.full_name.ilike(f"%{q}%"))

    query = query.order_by(User.created_at.desc())
    result = await db.execute(query)
    rows = result.all()

    return [{
        "id": user.id,
        "name": user.full_name,
        "email": user.email,
        "phone": user.phone or "",
        "age": profile.age if profile else None,
        "gender": profile.gender if profile else None,
        "blood_group": profile.blood_group if profile else None,
    } for user, profile in rows]


# ── Create appointment (assign doctor) ──────────────────
@router.post("/appointments")
async def create_appointment(
    data: ReceptionistAppointmentCreate,
    current_user: User = Depends(require_role(UserRole.RECEPTIONIST)),
    db: AsyncSession = Depends(get_db)
):
    # Get patient
    result = await db.execute(select(User).where(User.id == data.patient_id, User.role == UserRole.PATIENT))
    patient = result.scalar_one_or_none()
    if not patient:
        raise HTTPException(status_code=404, detail="Patient not found")

    # Get doctor
    result = await db.execute(select(User).where(User.id == data.doctor_id, User.role == UserRole.DOCTOR))
    doctor = result.scalar_one_or_none()
    if not doctor:
        raise HTTPException(status_code=404, detail="Doctor not found")

    # Get doctor profile for specialization
    result = await db.execute(select(DoctorProfile).where(DoctorProfile.user_id == data.doctor_id))
    doc_profile = result.scalar_one_or_none()

    appointment = Appointment(
        patient_id=data.patient_id,
        doctor_id=data.doctor_id,
        patient_name=patient.full_name,
        doctor_name=doctor.full_name,
        specialization=doc_profile.specialization if doc_profile else "",
        date=data.date,
        time=data.time,
        reason=data.reason,
    )
    db.add(appointment)
    await db.commit()
    await db.refresh(appointment)

    return {
        "message": f"Appointment booked — {patient.full_name} with {doctor.full_name}",
        "appointment_id": appointment.id
    }


# ── List all appointments ───────────────────────────────
@router.get("/appointments")
async def list_appointments(
    date_filter: str = "",
    current_user: User = Depends(require_role(UserRole.RECEPTIONIST)),
    db: AsyncSession = Depends(get_db)
):
    query = select(Appointment).order_by(Appointment.created_at.desc())

    if date_filter:
        query = query.where(Appointment.date == date_filter)

    result = await db.execute(query)
    appointments = result.scalars().all()

    return [{
        "id": a.id,
        "patient_name": a.patient_name,
        "doctor_name": a.doctor_name,
        "specialization": a.specialization,
        "date": a.date,
        "time": a.time,
        "status": a.status.value,
        "reason": a.reason or "",
    } for a in appointments]


# ── Notifications ───────────────────────────────────────
@router.get("/notifications")
async def get_notifications(
    current_user: User = Depends(require_role(UserRole.RECEPTIONIST)),
    db: AsyncSession = Depends(get_db)
):
    from datetime import date, timedelta
    from app.models.medicine import Medicine
    
    exp_date = (date.today() + timedelta(days=30)).isoformat()
    result = await db.execute(
        select(Medicine).where(
            Medicine.expiry_date <= exp_date,
            Medicine.expiry_date != "",
            Medicine.stock > 0
        )
    )
    expiring = [{
        "id": m.id,
        "name": m.name,
        "stock": m.stock,
        "expiry_date": m.expiry_date
    } for m in result.scalars().all()]
    
    return {"expiring_medicines": expiring}
